executor.py

```diff
@@ -1,7 +1,152 @@
+        self.traversal_paths = traversal_paths
+        self.logger = JinaLogger(
+            getattr(self.metas, 'name', self.__class__.__name__)
+        ).logger
+
     @requests
-    def foo(self, docs: DocumentArray, **kwargs):
-        pass
+    def encode(self, docs: DocumentArray, parameters: Dict = {}, **kwargs):
+        """
+        Read the numpy arrays from the Document.blob, and encode it into embeddings using the
+        comparable hashing technique. Features in the image are used to generate a distinct
+        (but not unique) fingerprint, and these fingerprints are comparable.
+
+        :param docs: A document array with documents to create embeddings for. Only the
+            documents that have the ``blob`` attribute will get embeddings. The ``blob``
+            attribute should be the numpy array of the image, and should have dtype
+            ``np.uint8``
+        :param parameters: A dictionary that contains parameters to control encoding.
+            The accepted keys are: `hash_type`, `hash_size`, `perceptual_hash_args`,
+            `average_hash_args`, `wavelet_hash_args`, and `traversal_paths`.
+            For example: `parameters={'hash_type': 'average', 'average_hash_args': {'avg_pixels': 'median'}}`
+        """
+        missing_blob = []
+        hash_type = parameters.get('hash_type', self.hash_type)
+        hash_size = parameters.get('hash_size', self.hash_size)
+
+        for doc in docs.traverse_flat(
+            parameters.get('traversal_paths', self.traversal_paths)
+        ):
+            if doc.blob is None:
+                missing_blob.append(doc.id)
+                continue
+
+            image = Image.fromarray(doc.blob)
+            hash_hex = None
+            try:
+                if hash_type == 'wavelet':
+                    wavelet_hash_args = deepcopy(self._wavelet_hash_args)
+                    wavelet_hash_args.update(parameters.get('wavelet_hash_args', {}))
+                    hash_hex = imagehash.whash(
+                        image, hash_size=hash_size, **wavelet_hash_args
+                    )
+                elif hash_type == 'average':
+                    average_hash_args = deepcopy(self._average_hash_args)
+                    average_hash_args.update(parameters.get('average_hash_args', {}))
+                    hash_hex = imagehash.average_hash(
+                        image, hash_size=hash_size, **average_hash_args
+                    )
+                elif hash_type == 'perceptual':
+                    perceptual_hash_args = deepcopy(self._perceptual_hash_args)
+                    perceptual_hash_args.update(
+                        parameters.get('perceptual_hash_args', {})
+                    )
+                    hash_hex = imagehash.phash(
+                        image, hash_size=hash_size, **perceptual_hash_args
+                    )
+                else:
+                    hash_hex = imagehash.phash(image, hash_size=hash_size)
```

Comment:
`difference` is using `dhash`, right?
```suggestion
                    hash_hex = imagehash.dhash(image, hash_size=hash_size)
```